### PR TITLE
blog: Zoekt announcement + tackling the long tail

### PR DESCRIPTION
**Pull request by stefanhengl**
DONT MERGE

This blog post motivates our efforts to add support for shard merging to Zoekt.

### CODE REVIEW

File: blogposts/2021/tackling-the-long-tail.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+. This file is called a shard. For small repositories, there is a 1:1 relationship between the repository and its shard.
+Larger repositories, such as Kubernetes map to several shards. At query time, shards are treated independently.
+
+## The long tail
+
+Most repositories are tiny and fit within a single shard. The following
+histogram shows the distribution of shard sizes on one of our production
+instances.
+
+![Distribution of shard sizes](https://storage.googleapis.com/sourcegraph-assets/blog/tackling-long-tail/tackling-long-tail-histogram.png)
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply blogposts/2021/tackling-the-long-tail.md:52]: The 100 MB relates to input data. The overhead of 2-3 times explains the larger shard sizes

>  "distribution of shard sizes" not "for"
Thanks! I will update it.

File: blogposts/2021/tackling-the-long-tail.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+## What is Zoekt?
+
+[Zoekt](https://github.com/sourcegraph/zoekt) is a code search engine originally created by Han-Wen Niehuys that
+performs trigram-based regex search. It is fast, easy to deploy and easy to maintain, which makes it a great choice for
+our self-hosted customer deployments.
+
+## Why not just scale out?
+
+Naturally, Zoekt's size of the index scales with the size of the input data. Zoekt’s index has an overhead of about 2-3
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply blogposts/2021/tackling-the-long-tail.md]: I updated the wording a bit, I hope it is clearer now.

File: blogposts/2021/tackling-the-long-tail.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+We can see that even tiny shards can have a lot of trigrams. As is to be
+expected, there is a positive correlation (the spearman correlation is 0.94) but
+the slope is very small. Effectively, we are paying a premium for small shards
+as they take up a lot of memory per byte of input data.
+
+The key insight is that there is a non-zero overlap between sets of trigrams
+from different shards. This presents a huge opportunity to save memory by
+merging small shards into much larger compound shards. The median
+intersection-over-union for two random shards in our sample is 0.13, which is a
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply blogposts/2021/tackling-the-long-tail.md]: I added a link to provide more backgroud and changed the formulation a bit. IoU is a normalized, dimensionless statistic, quantifying the overlap between 2 sets. 0 means no overlap, 1 means one set contains the other.

File: blogposts/2021/tackling-the-long-tail.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
+---
+title: 'How we’re tackling the long tail of tiny repos with shard merging'
+description: 'Sourcegraph is on track to grow its index of open source repositories significantly, with the aim of indexing the OSS universe. This post dives into the motivations behind introducing shard merging to our search backend.'
+author: Stefan Hengl
+publishDate: 2021-11-02T10:00-07:00
```
Comment (rebeccadee) [inline blogposts/2021/tackling-the-long-tail.md:5]:
```suggestion
publishDate: 2021-11-09T10:00-07:00
```

File: blogposts/2021/tackling-the-long-tail.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+In this post we give you a brief overview of how we're tackling the long tail of tiny repositories by introducing shard
+merging to Zoekt, in
+our [quest to index the OSS universe](https://about.sourcegraph.com/blog/why-index-the-oss-universe/).
+
+## What is Zoekt?
+
+[Zoekt](https://github.com/sourcegraph/zoekt) is a code search engine that
+performs trigram-based regex search. Originally created by Han-Wen Niehuys, Zoekt is fast, easy to deploy, and easy to maintain, which makes it a great choice for
+our self-hosted customer deployments. [Sourcegraph is actually taking on maintainership of Zoekt, which you can read about here](ADD LINK).
```
Comment (rebeccadee) [inline blogposts/2021/tackling-the-long-tail.md:23]: TODO

File: blogposts/2021/tackling-the-long-tail.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+Larger repositories, such as [Kubernetes](https://sourcegraph.com/github.com/kubernetes/kubernetes) or the [Linux Kernel](https://sourcegraph.com/github.com/torvalds/linux) map to several shards. At query time, shards are treated independently.
+
+## The long tail
+
+Most repositories are tiny and fit within a single shard. The following histogram shows the distribution of shard sizes
+on one of our production instances.
+
+<figure>
+  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/tackling-long-tail/tackling-long-tail-histogram.png" alt="Distribution of shard sizes" class="no-shadow">
+</figure>
```
Comment (rebeccadee) [inline blogposts/2021/tackling-the-long-tail.md:52]: @stefanhengl Just an FYI, I changed how the images are included so we can add captions and remove the drop shadow, which always looks a bit odd to me on white-background images!

**Comment by rebeccadee** [comment]
@stefanhengl Thanks so much for creating the PR for this. I've added the hero image and made some copy edits and formatting changes. I also tweaked the title and description to convey that this is an ongoing effort. If you have any questions/objections just let me know, otherwise this is almost ready to go for next week. I may add the Zoekt maintainership announcement to this PR so we can safely crosslink, so leave this with me for now!

**Comment by stefanhengl** [comment]
@rebeccadee We still have to update all links that point to code after we have completed the transfer of Zoekt. The links point to our fork and will break depending on how the transfer is done.

**Comment by stefanhengl** [comment]
@rebeccadee I just checked. Keegan already unforked our Zoekt fork and the links are still working. @jjeffwarner will reach out soon to time the updates of the respective READMEs.

**Comment by jjeffwarner** [comment]
@rebeccadee We are in a good spot on the repo. We are ready to move forward on this whenever works for you.

**Comment by rebeccadee** [comment]
> @rebeccadee We are in a good spot on the repo. We are ready to move forward on this whenever works for you.

@jjeffwarner Thank you! We've scheduled the announcement for tomorrow so I will merge then 🎉

**Comment by jjeffwarner** [comment]
Perfect!

**Pull request merged**